File: engines/julia/benchmark.jl
EXIT_FAILURE = 1

function measure(data, pattern)

    try
        compiled_pattern = Regex(pattern)
        start = time()
        count = length(collect(eachmatch(compiled_pattern, data)))
        elapsed = time() - start
        elapsed_ms = 1000 * elapsed

        println(string(elapsed_ms) * " - " * string(count))
    catch e
        println(stderr, "compilation error: " * e)
        exit(EXIT_FAILURE)
    end

end

if length(ARGS) != 3
    println("Usage: julia benchmark.jl <filename> regex num_iterations")
    exit(EXIT_FAILURE)
end

data = open(f->read(f, String), ARGS[1])
pattern = ARGS[2]
num_iterations = parse(Int64, ARGS[3])
for i in 1:num_iterations
    measure(data, pattern)
end

File: engines/ruby/benchmark.rb
require "benchmark"

EXIT_FAILURE = 1

if(ARGV.size != 3)
  puts "Usage: ruby benchmark.rb <filename> regex numIterations"
  exit EXIT_FAILURE
end

def measure(data, pattern)
  begin
    count = 0
    regex = Regexp.compile(pattern)
  
    elapsed = Benchmark.measure {
      count = data.scan(regex).size
    }
  
    puts "#{elapsed.real * 1e3} - #{count}"
  rescue => error
    STDERR.puts "compilation error: #{error}\n"
    exit EXIT_FAILURE
  end
end

data = File.read(ARGV[0])
pattern = ARGV[1]
numIterations = Integer(ARGV[2])

for i in 0...numIterations
  measure(data, pattern)
end
